Fixed race condition, and updated async code.

aronsky opened the pull request:
## Description:

The main issue fixed in this pull request is a race condition between the keep-alive mechanism and the tracking of temperature.

As a refresher on the inner workings of the generic thermostat, it tracks temperature and signals the A/C motor/heater to turn on when the temperature goes above (when in cooling mode)/below (when in heating mode) a given threshold. When the opposite happens, it signals the motor/heater to turn off. Meanwhile, a keep-alive option retransmits the turn on/off signal, based on the current thermostat's state. It is repeating via `async_track_time_interval`.

When the keep-alive event fires right before a threshold-crossing temperature change event, a race condition appears. The keep-alive code will attempt to retransmit the signal for the current state of the thermostat (e.g. turn on, when the device is active). However, the temperature change event will attempt to do the opposite (e.g. turn it off if it was active and the temperature threshold was reached). If the temperature change is handled faster, the A/C will be turned off, and immediately afterwards turned on by the slower keep-alive handler. At this point, not only the A/C is on when it should be off - but it'll stay that way for the minimum duration cycle, therefore following keep-alive events and temperature changes won't be able to remedy the issue.

The issue is fixed via a locking mechanism. At the same time, I took the liberty of refreshing and refactoring the code, making most of it `async`, and simplifying some functions. Additionally, I renamed the heater-related names to ac/ac_motor, as that's more generic (the thermostat is not intended for use with heaters only, after all).

**Related issue (if applicable):** fixes #<home-assistant issue number goes here>

**Pull request in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io) with documentation (if applicable):** home-assistant/home-assistant.github.io#<home-assistant.github.io PR number goes here>

## Example entry for `configuration.yaml` (if applicable):
```yaml

```

## Checklist:
  - [X] The code change is tested and works locally.
  - [X] Local tests pass with `tox`. **Your PR cannot be merged unless tests pass**

If user exposed functionality or configuration variables are added/changed:
  - [ ] Documentation added/updated in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io)

If the code communicates with devices, web services, or third-party tools:
  - [ ] New dependencies have been added to the `REQUIREMENTS` variable ([example][ex-requir]).
  - [ ] New dependencies are only imported inside functions that use them ([example][ex-import]).
  - [ ] New or updated dependencies have been added to `requirements_all.txt` by running `script/gen_requirements_all.py`.
  - [ ] New files were added to `.coveragerc`.

If the code does not interact with devices:
  - [ ] Tests have been added to verify that the new code works.

[ex-requir]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L14
[ex-import]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L54

balloob committed:
Bumped version to 0.74.0b0

balloob committed:
0.73.2 - security release (#15494)

* Extract SSL context creation to helper (#15483)

* Extract SSL context creation to helper

* Lint

* Bumped version to 0.73.2

balloob committed:
Merge branch 'master' into rc

balloob committed:
Bumped version to 0.74.0b1

balloob committed:
Add current user WS command (#15485)

mjg59 committed:
Update HomeKit module code (#15502)

This fixes a bunch of bugs, including issues with concurrency in devices
that present multiple accessories, devices that insist on the TLV entries
being in the order that Apple use, and handling devices that send headers
and data in separate chunks. This should improve compatibility with
a whole bunch of HomeKit devices.

balloob committed:
Add onboarding support (#15492)

* Add onboarding support

* Lint

* Address comments

* Mark user step as done if owner user already created

balloob committed:
Bump frontend to 20180717.0

balloob committed:
Bumped version to 0.74.0b2

awarecan committed:
Disallow use insecure_example auth provider in configuration.yml (#15504)

* Disallow use insecure_example auth provider in configuration.yml

* Add unit test for auth provider config validate

aronsky committed:
Fixed race condition, and updated async code.

aronsky committed:
Fixed race condition, and updated async code.

MartinHjelmare commented:
It's hard to follow the changes in this PR. Please limit a PR to one fix or one feature. So the fix should go in one PR, the refactor in another and the rename in a third.

scop committed:
Remove some unused imports (#15529)

balloob committed:
Decouple emulated hue from http server (#15530)

balloob committed:
Decouple emulated hue from http server (#15530)

pvizeli committed:
Moon translate (#15498)

* Translate moon

* Create strings.moon.json

* Update moon.py

* Update strings.moon.json

* Update test_moon.py

scop committed:
More pylint 2 fixes (#15516)

* Pylint 2 useless-import-alias fixes

* Pylint 2 chained-comparison fixes

* Pylint 2 consider-using-get fixes

* Pylint 2 len-as-condition fixes

andrey-git committed:
Make sure that only pypi dependencies are used (#15490)

amelchio committed:
RFC: Only use supported light properties (#15484)

* Only use supported light properties

* Fix tests

worm-ee committed:
Cleanup of HomematicIP Cloud code (#15475)

* Check if device supports lowBat and shows it only if battery is low

* Show empty battery icon if lowBat is true

* Default return None

* Sabotage attribute and icon if device has this feature

* Bug fix and cleanup

* Use dedicated function for security state

* Cleanup of sensor attributes and icons

* Empty

dshokouhi committed:
Implement locate service for neato (#15467)

* Implement locate service for neato

* Hound

turbokongen committed:
Make RS room thermostat discoverable (#15451)

* Make RS room thermostat discoverable

* Reversed generic type name

aronsky commented:
@MartinHjelmare, I understand, and will try to work in a more structured way in the future. However, in this PR, several `async` upgrades of methods are required by the race condition fix - submitting them separately wouldn't make a lot of sense, and would result in code that's styled inconsistently. Furthermore, splitting this PR across 3 different ones would slow down the propagation, as each of them would rely on the merging of the previous one.

fucm committed:
Add support for Tahoma Soke Sensor (#15441)

teharris1 committed:
Mini-Remote events (#15523)

* Add event handler to capture binary sensor on messages

* Log event trigger

* Log event firing

* Capture platform correctly

* Fix test for platform eq binary_sensor

* Create sensor events

* Add light and battery sensors

* Bump insteonplm version to 0.11.6

* Fix naming of BUTTON_PRESSED_STATE_NAME

* Fix naming of fire event methods

* Add logging

* Add DOMAIN definition

* Get state name from plm.devices

* Remove stale reference to button ID

* Fix reference to state name

* Remove incorrect ref to self

* Log remote button pressed event

* Change mode to button_mode and fix values to array

* Rename CONF_MODE to CONF_BUTTON_MODE

* Log platform create with mode

* Properly assign button_mode to track mode

* Implement is_on

* Change mini-remotes to events only

* Remove button_mode config option

* Fix reference to _fire_button_on_off_event

* Bump insteon version to 0.11.7

* Flake8 clean up

* Flake8 cleanup

* Use % format in logging per pylint

* Code review updates

* Resolve conflict

* Lint

gieljnssns committed:
Update pyatmo (#15540)

balloob committed:
Bump frontend to 20180718.0

balloob committed:
Bump frontend to 20180718.0

balloob committed:
Bumped version to 0.74.0b3

balloob committed:
Don't be so strict client-side (#15546)

balloob committed:
Don't be so strict client-side (#15546)

balloob commented:
I disagree that it would slow down the propagation. Small PRs are very easy to review and move very fast. We have a `small-pr` label for a reason :)

aronsky commented:
Maybe, but like I said, in this case - most of the PR is the actual fix of the race condition. Splitting it into 3 PRs now is counter-productive - it will require me to redo the work from scratch, and eventually will result in 1 PR that's almost as big as the current one (for the fix), and then additionally 2 smaller PRs.

Like I said - I will make the PRs more structured in the future, but at this point redoing it from scratch makes no sense. Furthermore, on top of personally verifying that the code works properly (I know, it's just my personal experience, and doesn't cover all possible configurations, but it's better than nothing), the end result is actually more readable than the original code, and certainly makes more sense than any in-between progress.

I suggest to invite people who previously worked on the generic thermostat to review it, and if the changes still look out of context - to review the end result itself, instead. Alternatively, I'd rather comment the changes to make the review easier, than starting this whole process from scratch when a perfectly working solution is already in place.

quthla committed:
Update reading when device is added (#15548)

balloob committed:
Allow CORS requests to token endpoint (#15519)

* Allow CORS requests to token endpoint

* Tests

* Fuck emulated hue

* Clean up

* Only cors existing methods

balloob committed:
Allow CORS requests to token endpoint (#15519)

* Allow CORS requests to token endpoint

* Tests

* Fuck emulated hue

* Clean up

* Only cors existing methods

balloob committed:
Update config entry id in entity registry (#15531)

jeradM committed:
Added WS endpoint for changing homeassistant password. (#15527)

* Added WS endpoint for changing homeassistant password.

* Remove change password helper. Don't require current password.

* Restore current password verification.

* Added tests.

* Use correct send method

jeradM committed:
Added WS endpoint for changing homeassistant password. (#15527)

* Added WS endpoint for changing homeassistant password.

* Remove change password helper. Don't require current password.

* Restore current password verification.

* Added tests.

* Use correct send method

balloob committed:
Bump frontend to 20180719.0

balloob committed:
Bump frontend to 20180719.0

balloob committed:
Bumped version to 0.74.0b4

Commit:
Upgrade to simplisafe-python v2 to use new SimpliSafe API (#15542)

* Upgrade to simplisafe-python v2 to use new SimpliSafe API

balloob committed:
Allow auth providers to influence is_active (#15557)

* Allow auth providers to influence is_active

* Fix auth script test

balloob committed:
Allow auth providers to influence is_active (#15557)

* Allow auth providers to influence is_active

* Fix auth script test

MartinHjelmare committed:
Upgrade pymysensors to 0.16.0 (#15554)

MartinHjelmare committed:
Upgrade pymysensors to 0.16.0 (#15554)

balloob committed:
Remove relative time from state machine (#15560)

balloob committed:
Remove relative time from state machine (#15560)

scop committed:
More pylint 2 fixes (#15565)

## Description:

More fixes flagged by pylint 2 that don't hurt to have before the actual pylint 2 upgrade (which I'll submit soon).

## Checklist:
  - [ ] The code change is tested and works locally.
  - [x] Local tests pass with `tox`. **Your PR cannot be merged unless tests pass**

huangyupeng committed:
Add Tuya fan support (#15525)

* Add Tuya fan platform

* Add Tuya fan platform

* fix as review required

awarecan committed:
Reset failed login attempts counter when login success (#15564)

awarecan committed:
Reset failed login attempts counter when login success (#15564)

balloob committed:
Update the frontend to 20180720.0

balloob committed:
Update the frontend to 20180720.0

balloob committed:
Bumped version to 0.74.0

rytilahti committed:
light.tplink: initialize min & max mireds only once, avoid i/o outside update (#15571)

* light.tplink: initialize min & max mireds only once, avoid i/o outside update

* revert the index change

* fix indent, sorry for overwriting your fix, balloob

rytilahti committed:
light.tplink: initialize min & max mireds only once, avoid i/o outside update (#15571)

* light.tplink: initialize min & max mireds only once, avoid i/o outside update

* revert the index change

* fix indent, sorry for overwriting your fix, balloob

balloob committed:
Merge pull request #15570 from home-assistant/rc

0.74

azogue committed:
fix aiohttp InvalidURL exception when fetching media player image (#15572)

* fix aiohttp InvalidURL exception when fetching media player image

The first call for the HA proxy (`/api/media_player_proxy/media_player.kodi?token=...&cache=...`)
is receiving relative urls that are failing, this is a simple fix to precede the base_url when hostname is None.

* fix import location and sort stdlib imports

balloob committed:
Merge branch 'master' into dev

PrimusNZ committed:
Allow MQTT Switch to have an optional state configuration (#15430)

Switches by default use the payload_on and payload_off configuration parameters to specify both the payload the switch should send for a state but also what will be returned for the current state - which isnt always the same
As a toggle switch might always send an ON or TOGGLE to toggle the switch, but still receive an ON or an OFF for the state topic - This change allows for splitting them apart

amelchio commented on homeassistant/components/climate/generic_thermostat.py:
```diff
@@ -278,33 +278,23 @@ def max_temp(self):
 
-    @callback
-    def _async_update_temp(self, state):
+    async def _async_update_temp(self, state):
```
Why is this now async?

aronsky replied:
It works as either `async` or `callback` as it used to be. I marked the methods that can be asynchronous as `async` in order to:
* make the code more consistent (avoid mixing async and regular code) and enforce its asynchronicity in the future
* make it consistent with the styling of having all the functions/methods named `async_???` be actually asynchronous - this way, people working on the code will see a method name that starts with `async`, and will be able to safely assume that they can `await` on that method.

amelchio replied:
I understand the desire to not mix styles. Checking this again, though, it appears that you forgot to add `await` to a call on line 147. Doing so will not be easy if you want to maintain the current consistency. So you might want to reconsider :)

Note that the usual Home Assistant convention is that functions prefixed with `async_` must be run in the event loop (i.e. not in worker threads). It does _not_ mean they are necessarily `async def`. So it is fine to name callbacks `async_`, even if it is confusing.

amelchio reviewed (commented):
You are asking reviewers to spend some extra time because you cannot be bothered yourself. That is going to make reviewers feel disrespected and thus less likely to help you out.

I did review your changes though and they look fine.

I am not sure that I like the rename. You say "air conditioning" is more generic but to me it is also more ambiguous. Especially since there is an `ac_mode` where "ac" means cooling. Maybe "device" or even "switch" could work?

amelchio committed:
Add unique_id to netgear_lte sensors (#15584)

aronsky commented:
@amelchio, I understand the sentiment, and I appreciate the time you took to go over the changes.

I agree with you regarding the ambiguity of the `air conditioning` name, given the existence of `ac_mode`. I was looking for a word that describes the device to be toggled, and A/C sounded more relevant than heater, as A/C can be used both for heating and cooling purposes. `device` or `switch` might sound too generic - how about `climate device`? I plan on doing more work on this component, and further improve the support for both heating and cooling in a single configuration, at which point `ac_mode` will be probably replaced by something more appropriate (like `mode` or `features` that supports options like `auto`, `cool_only`, and `heat_only`). But that's definitely beyond the scope of this PR. The renaming was supposed to be part of the groundwork for the future.

digiblur committed:
Add relay addr & chan config to alarmdecoder zones (#15242)

Add relay addr & chan config to alarmdecoder zones

danielperna84 committed:
Update pyhomematic, fixes #15054, #15190 (#15603)

amelchio committed:
Throttle unavailability warnings for tplink light/switch (#15591)

awarecan committed:
Frontend component should auto load auth coomponent (#15606)

cdce8p committed:
Bugfix HomeKit name and serial_number (#15600)

* Bugfix HomeKit name and serial_number

* Revert serial_number changes

balloob committed:
Remove entity picture of Tuya entity (#15611)

scop committed:
Upgrade mypy to 0.620 (#15612)

scop committed:
Pylint cleanups (#15626)

* Pylint 2 no-else-return fixes

* Remove unneeded abstract-class-not-used pylint disable

andrey-git committed:
Add typing to homeassistant/*.py and homeassistant/util/ (#15569)

* Add typing to homeassistant/*.py and homeassistant/util/

* Fix wrong merge

* Restore iterable in OrderedSet

* Fix tests

amelchio committed:
Use case insensitive comparison for Sonos model check (#15604)

amelchio committed:
Refresh Sonos source list on changes (#15605)

sherazlodhi committed:
The tense is wrong (#15614)

Instead of spent, we should be using spend :)

starkillerOG committed:
Suppress error between 00:00 and 01:00 (#15555)

* Suppress error between 00:00 and 01:00

Suppress an error that often occers between 00:00 and 01:00 CE(S)T during that time, probably because buienradar.nl is then updating its forcast for the next day. The API does not always work between these times (in the middle of the night).

* white space & import

* unnecessary brackets

balloob committed:
Use async_create_task (#15633)

* Use async_create_task

* Fix test

balloob committed:
Allow system users to refresh tokens (#15574)

balloob committed:
WIP: Hass.io sent token to supervisor (#15536)

Hass.io sent token to supervisor

pvizeli committed:
Fix aiohttp connection reset errors (#15577)

* Fix aiohttp connection reset errors

* Update aiohttp_client.py

* Update aiohttp_client.py

* Update __init__.py

* Update mjpeg.py

* Update mjpeg.py

* Update ffmpeg.py

* Update ffmpeg.py

* Update ffmpeg.py

* Update proxy.py

* Update __init__.py

* Update aiohttp_client.py

* Update aiohttp_client.py

* Update proxy.py

* Update proxy.py

* Fix await inside coroutine

* Fix async syntax

* Lint

balloob committed:
Cast/Sonos: create config entry if manually configured (#15630)

* Cast/Sonos: create config entry if manually configured

* Add test for helper

turbokongen committed:
Add ability to set Zwave protection commandclass (#15390)

* Add API for protection commandclass

* Adjusting

* tests

* Spelling

* Missed flake8

* Period

* spelling

* Review changes

* removing additional .keys()

* period

* Move i/o out into executor pool

* Move i/o out into executor pool

* Forgot get method

* Do it right... I feel stupid

* Long lines

* Merging

dshokouhi committed:
Update Neato library to allow for dynamic endpoints (#15639)

wingy3181 committed:
Fix typo `asayn_init` instead of `async_init` (#15645)

awarecan committed:
Add ipban for failed login attempt in new login flow (#15551)

* Add ipban for failed login attempt in new login flow

* Address review comment

* Use decorator to clean up code

huangyupeng committed:
Add Tuya cover and scene platform (#15587)

* Add Tuya cover platform

* Add Tuya cover and scene

* fix description

* remove scene default method

visibilityspots committed:
Update smappy library version (#15636)

Adding latest smappy lib version

Updated smappy library version

balloob committed:
Allow changing entity ID (#15637)

* Allow changing entity ID

* Add support to websocket command

* Address comments

* Error handling

balloob committed:
Bump frontend to 20180724.0

glpatcern committed:
Downgrade netatmo warning log to info (#15652)

awarecan committed:
Add turn_on/off service to camera (#15051)

* Add turn_on/off to camera

* Add turn_on/off supported features to camera.

Add turn_on/off service implementation to camera, add turn_on/off
 supported features and services to Demo camera.

* Add camera supported_features tests

* Resolve code review comment

* Fix unit test

* Use async_add_executor_job

* Address review comment, change DemoCamera to local push

* Rewrite tests/components/camera/test_demo

* raise HTTPError instead return response

kalmard0 committed:
Allow defining default turn-on values for lights in the profiles file. (#15493)

* Allow defining default turn-on values for lights in the profiles file.

* Mock out file operations in unit test.

* Fix unit test flakiness.

* Avoid unnecessary copy

jeradM committed:
Added user credentials to current_user ws endpoint. (#15558)

* Added user credentials to current_user ws endpoint.

* Comments. Added another test.

* Return list of credentials.

scop committed:
Import collections abstract base classes from collections.abc (#15649)

Accessing them directly through collections is deprecated since 3.7, and
will no longer work in 3.8.

scop committed:
Auth typing improvements (#15640)

* Always return bytes from auth.providers.homeassistant.hash_password

Good for interface cleanliness, typing etc.

* Add some homeassistant auth provider type annotations

balloob committed:
Fix CORS duplicate registration (#15670)

peternijssen committed:
Add spider thermostat (#15499)

* add spider thermostats

* Added load_platform. Added operation dictionary. Minor improvements

* loop over spider components for load_platform

* added empty dict to load_platform. changed add_devices

* moved logic to the API

* fix requirements_all.txt

* minor code improvements

eavanvalkenburg committed:
Add Brunt Cover Device (#15653)

* New Brunt Branch

* Some small changes and updates based on review.

scop committed:
Pylint 2 useless-return fixes (#15677)

awarecan committed:
Allow Nest Cam turn on/off (#15681)

* Allow Nest Cam turn on/off

* Don't raise Error

* Remove unnecessary state update

scop committed:
Upgrade pylint to 2.0.1 (#15683)

* Upgrade pylint to 2.0.1

* Pylint 2 bad-whitespace fix

* Pylint 2 possibly-unused-variable fixes

* Pylint 2 try-except-raise fixes

* Disable pylint fixme for todoist for now

https://github.com/PyCQA/pylint/pull/2320

* Disable pylint 2 useless-return for now

https://github.com/PyCQA/pylint/issues/2300

* Disable pylint 2 invalid-name for type variables for now

https://github.com/PyCQA/pylint/issues/1290

* Disable pylint 2 not-an-iterable for now

https://github.com/PyCQA/pylint/issues/2311

* Pylint 2 unsubscriptable-object workarounds

* Disable intentional pylint 2 assignment-from-nones

* Disable pylint 2 unsupported-membership-test apparent false positives

* Disable pylint 2 assignment-from-no-return apparent false positives

* Disable pylint 2 comparison-with-callable false positives

https://github.com/PyCQA/pylint/issues/2306

balloob committed:
Bump frontend to 20180726.0

balloob committed:
Fix frontend tests

bachya committed:
Fixes a bug with showing a subset of Pollen index conditions (#15694)

peternijssen committed:
Add spider power plug component (#15682)

* Add spider power plug component

* rounding down the numbers

* ability to throttle the API

* updated to the lastest api

* resolved an issue within the API

juhaniemi committed:
Add support for P5 FutureNow light platform (#15662)

* Added support for FutureNow light platform and relay/dimmer units

* Pinned specific version for requirement

* Added support for FutureNow light platform and relay/dimmer units

* Added futurenow.py to .coveragerc.

* Minor fixes and enhancements as requested in the code review.

* Minor fixes and enhancements as requested in the code review.

* Use device_config's value directly as it's validated as boolean.

* Simplify state check.

* Fixed brightness update that was broken in previous commit.

awarecan committed:
Only log change to use access token warning once (#15690)

rorr73 committed:
Add support for alarm_control_panel to MQTT Discovery. (#15689)

jcconnell committed:
Add Magicseaweed API support (#15132)

* Added support for magicseaweed surf forecasting

* Added support for magicseaweed surf forecasting

* Added support for magicseaweed surf forecasting

* Incorporate @bachya requested changes.

* Adding support for magicseaweed package.

* Run tests and fix errors.

* Incorporate @balloob requested changes.

* Attempt to fix pylint error e1101.

* Two spaces before inline comments

* Add @MartinHjelmare & @balloob requested changes.

* Remove MagicSeaweedData object inheritance.

* Fix variable logic.

alanfischer committed:
Upgrade pyvera to 0.2.44 (#15708)

dingusdk committed:
Remove IHC XML Element from discovery data (#15719)

balloob committed:
Context (#15674)

* Add context

* Add context to switch/light services

* Test set_state API

* Lint

* Fix tests

* Do not include context yet in comparison

* Do not pass in loop

* Fix Z-Wave tests

* Add websocket test without user

awarecan committed:
Break up components/auth (#15713)

jkeljo committed:
Add a component for Sisyphus Kinetic Art Tables (#14472)

* Add a component for Sisyphus Kinetic Art Tables

The [Sisyphus Kinetic Art Table](https://sisyphus-industries.com/) uses a
steel ball to draw intricate patterns in sand, thrown into sharp relief by a
ring of LED lights around the outside.

This component enables basic control of these tables through Home Assistant.

* Fix lints

* Docstrings, other lints

* More lints

* Yet more.

* Feedback

* Lint

* Missed one piece of feedback

* - Use async_added_to_hass in media player
- async_schedule_update_ha_state in listeners
- constants for supported features
- subscripting for required keys
- asyncio.wait
- update to sisyphus-control with passed-in session

* Update requirements

* lint

Commit:
Command Line Sensor - json_attributes (#15679)

* Add tests to command_line for json_attrs

* Add json_attrs to command_line

* Remove whitespace on blank line

* Stick to <80 row length

* Use collections.Mapping, not dict

* Rename *attrs to *attributes

* Remove extraneous + for string concat

* Test multiple keys

* Add test

Makes sure the sensor's attributes don't contain a value for a missing key,
even if we want that key.

* Test that unwanted keys are skipped

* Remove additional log line

* Update tests for log changes

* Fix ordering

fabaff committed:
Upgrade youtube_dl to 2018.07.21 (#15718)

fabaff committed:
Upgrade numpy to 1.15.0 (#15722)

peternijssen committed:
Upgrade spiderpy to 1.2.0 (#15729)

fabaff committed:
Upgrade beautifulsoup4 to 4.6.1 (#15727)

fabaff committed:
Upgrade mutagen to 1.41.0 (#15739)

fabaff committed:
Upgrade sqlalchemy to 1.2.10 (#15737)

fabaff committed:
Upgrade voluptuous to 0.11.3 (#15735)

fabaff committed:
Upgrade youtube_dl to 2018.07.29 (#15734)

fabaff committed:
Upgrade pyowm to 2.9.0 (#15736)

fabaff committed:
Various updates (#15738)

starkillerOG committed:
Denonavr version push to 0.7.5 (#15743)

* Version push to 0.7.5

Improve logger warning

* Denonavr v.0.7.5

juhaniemi committed:
Make FutureNow light remember last brightness when turning on (#15733)

* Remember last brightness value and use it on turn_on()

* Pyfnip-0.2 now returns state reliably, no manual changes needed.

* Split too long line of code

* Updated pyfnip library version

shoejosh committed:
Add Genie Aladdin Connect cover component (#15699)

* Add Genie Aladdin Connect cover component

* Fix lines being too long

* Fix issues found in review

* remove Unknown state, use None instead

* Fixed requirements_all

amelchio commented:
The problem with "air conditioning" is that it means different things to different people. I believe that to many people living in hot places it just means cooling. Please revert the renaming and tackle that in a separate PR so it can get some discussion. This should be a simple search/replace job and not take much time.

The refactor would have been easier to review as an initial PR but I agree it is mostly a result of having to `await` under the new lock so you can keep that.

DavidMStraub committed:
Add mvglive option to store multiple departures in attributes (#15454)

* MVG Live sensor: add option to store multiple departures in attributes

* Fix lint error

* mvglive: take into account timeoffset in API call

* Prevent exception if departure list is empty

* Rename state_attributes -> device_state_attributes

danielfaulknor committed:
Add other wemo motion sensor identifier (#15627)

* Add other motion sensor identifier

* Fix order

awarecan committed:
Remove self type hints (#15732)

* Remove self type hints

* Lint

superpuffin committed:
Upgrade Adafruit-DHT to 1.3.3 (#15706)

* Change to newer pip package

The package Adafruit_Python_DHT==1.3.2 was broken and would not install, breaking DHT sensor support in Home assistant. It has since been fixed in Adafruit-DHT==1.3.3.

See: https://github.com/adafruit/Adafruit_Python_DHT/issues/99

* Update requirements_all.txt

New or updated dependencies have been added to `requirements_all.txt` by running `script/gen_requirements_all.py`.

* Comment out Adafruit-DHT

Adafruit_Python_DHT changed name to Adafruit-DHT, which still need pyx support breaking our CI, need to be comment out.

* Update requirements_all.txt

rytilahti committed:
Add kodi unique id based on discovery (#15093)

* kodi add unique id based on discovery

* initialize unique_id to None

* use netdisco-extracted mac_address

* use an uuid instead of mac for real uniqueness

* add missing docstring

* verify that there is no entity already for the given unique id

* whitespace fix

bachya committed:
Better handling of Yi camera being disconnected (#15754)

* Better handling of Yi camera being disconnected

* Handling video processing as well

* Cleanup

* Member-requested changes

* Member-requested changes

awarecan committed:
Decouple login flow view and data entry flow view (#15715)

foxel committed:
Fixed NDMS for latest firmware (#15511)

* Fixed NDMS for latest firmware.
Now using telnet instead of Web Interface

* Using external library for NDMS interactions

* updated requirements_all

* renamed `mac` to `device` back

* Using generators for name and attributes fetching

andrey-git committed:
Add disallow_untyped_calls to mypy check. (#15661)

* Add disallow_untyped_calls to mypy check.

* Fix generator

turbokongen committed:
Update pyozw to 0.4.9 (#15758)

* update pyozw to 0.4.8

* add requirements_all.txt

* use 0.4.9

fabaff committed:
Convert wind speed to km/h (fixes #15710) (#15740)

* Convert wind speed to km/h (fixes #15710)

* Round speed

priiduonu committed:
Round precipitation forecast to 1 decimal place (#15759)

The OWM returns precipitation forecast values as they are submitted to their network. It can lead to values like `0.0025000000000004 mm` which does not make sense and messes up the display card. This PR rounds the value to 1 decimal place.

salbertson committed:
Add a "Reviewed by Hound" badge (#15767)

fabaff committed:
Upgrade Mastodon.py to 1.3.1 (#15766)

dgomes committed:
This component API has been decomissioned on the 31st of May 2018 by Telstra (#15757)

See #15668

fabaff committed:
Fix docstrings (#15770)

MatMaul committed:
Update pynetgear to 0.4.1 (bugfixes) (#15768)

miawgogo committed:
Opensky altitude (#15273)

* Added Altitude to opensky

* decided to take all metadata

* Final Tidy

* More formatting

* moving CONF_ALTITUDE to platform

* Moved CONF_ALTITUDE to platform

soraxas committed:
Expose internal states and fixed on/off state of Dyson Fans (#15716)

* exposing internal state and fixed onoff state

* fixed styling

* revert file mode changes

* removed self type hints

* better unit test and changed the way to return attributes

* made wolfie happy

Kane610 committed:
deCONZ - support for power plugs (#15752)

* Initial commit for deCONZ switch support

* Fix hound comment

* Fix martins comment; platforms shouldn't depend on another platform

* Fix existing tests

* New tests

* Clean up unnecessary methods

* Bump requirement to v43

* Added device state attributes to light

cnrd committed:
Add support for STATES of vacuums (#15573)

* Vacuum: Added support for STATES

* Added debug logging and corrected state order

* typo

* Fix travis error, STATE = STATE for readability

* status -> state

* Changed to Entity instead of ToogleEntity

* Updated some vacuums

* Revert changes

* Revert Changes

* added SUPPORT_STATE

* Woof?

* Implement on/off if STATE not supported

* Moved new state vaccum to Class StateVacuumDevice

* Error: I should go to bed

* Moved around methods for easier reading

* Added StateVacuumDevice demo vacuum

* Added tests for StateVacuumDevice demo vacuum

* Fix styling errors

* Refactored to BaseVaccum

* Vacuum will now go back to dock

* Class BaseVacuum is for internal use only

* return -> await

* return -> await

niklaswa committed:
Add support for STATE_AUTO of generic_thermostat (#15678)

Add support for STATE_AUTO of generic_thermostat

awarecan committed:
Revert "Add support for STATE_AUTO of generic_thermostat (#15678)" (#15783)

This reverts commit 2e5131bb2161b43461231a683e45d2c843d727f0.

aronsky committed:
Merge branch 'generic-thermostat-fix-race-condition' of github.com:aronsky/home-assistant into generic-thermostat-fix-race-condition

aronsky commented:
And... I made a mistake with updating the Pull Request again :(
I am going to close it and redo it again - and since I'm redoing it anyway, I'll do it properly, in several stages.

Thank you for the review.
